doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,264 @@
+the secrets will be updated atomically.
+
+Note that while the catalog updates to a secret apply immediately, changes to
+the contents of the secret do not propagate to downstream consumers of the
+secret on any particular timeline.
+
+The `SHOW SECRETS` statement lists the *names* of the secrets in the specified
+schema, or the first schema in the search path if no schema is explicitly
+specified, optionally filtered by the provided `LIKE` pattern or `WHERE`
+expression. The filter clauses which work analogously to the same clauses in the
+`SHOW DATABASES` statement.
+
+Note that there is intentionally no way to view the contents of secrets.
+
+### System catalog changes
+
+A new `mz_secrets` table with the following structure will describe the
+available secrets in the system:
+
+Field       | Type      | Meaning
+------------|-----------|--------
+`id`        | `bigint`  | The ID of the secret.
+`schema_id` | `bigint`  | The ID of the schema to which the secret belongs.
+`name`      | `text`    | The name of the secret.
+
+Note that, again, the contents of the secrets are not exposed.
+
+### Referencing secrets
+
+There is presently no way to use a secret. The existing `CREATE SOURCE` and
+`CREATE SINK` commands will be extended to support referencing secrets, but
+that work is left to a future design document.
+
+## Implementation
+
+### Catalog changes
+
+The [`CatalogItem`] enum will be extended with an additional `Secret` variant:
+
+```rust
+pub enum CatalogItem {
+    // ...
+    Secret,
+}
+```
+
+Unlike the other catalog items, there is no data associated with a secret. The
+name and ID of the secret are stored in the `CatalogEntry` struct, and the
+actual data for the secret will be stored in a secret backend (see next
+section). Therefore there is presently no additional metadata that needs to be
+recorded about a secret.
+
+When serializing a secret, its `create_sql` will be synthesized as:
+
+```sql
+CREATE SECRET <name> AS '******'
+```
+
+When deserializing a secret on bootstrap, the `'******'` will be ignored.
+
+### Secrets controller
+
+The `Coordinator` will host a `SecretsController` that securely handles the
+storage of secrets. It will interact with the controller via the following
+interface.
+
+```rust
+/// Securely stores secrets.
+pub trait SecretsController {
+    /// Applies the specified secret operations in bulk.
+    ///
+    /// Implementations must apply the operations atomically. If the method
+    /// returns `Ok(())`, then all operations have been applied successfully;
+    /// if the method returns `Err(())`, then none of the operations have been
+    /// applied.
+    ///
+    /// Implementations are permitted to reject combinations of operations which
+    /// they cannot apply atomically.
+    fn apply(&mut self, ops: Vec<SecretOp>) -> Result<(), anyhow::Error>;
+
+    /// Returns the IDs of all known secrets.
+    fn list(&self) -> Result<Vec<GlobalId>, anyhow::Error>;
```

Comment:
for my clarity. Does this return all the secrets in the catalog, or all the secrets in the secret backend (K8s/FS)? How does this interact with the proposed garbage collection process?

Reply:
All the secrets in the secret controller (so K8s or filesystem). The idea is that this is how the garbage collection process discovers secrets that are orphaned in the controller.